crates/qpy/src/circuit_writer.rs

```diff
@@ -0,0 +1,910 @@
+            }
+            if op
+                .operation
+                .bind(py)
+                .is_instance(imports::ANNOTATED_OPERATION.get_bound(py))?
+            {
+                let op = op.operation.bind(py);
+                let modifiers = op.getattr("modifiers")?;
+                return modifiers
+                    .try_iter()?
+                    .map(|modifier| py_pack_param(&modifier?, qpy_data, Endian::Little))
+                    .collect::<PyResult<_>>();
+            }
+            if op
+                .operation
+                .bind(py)
+                .is_instance(imports::PAULI_PRODUCT_MEASUREMENT.get_bound(py))?
+            {
+                let op = op.operation.bind(py);
+                let pauli_data = op.call_method0("_to_pauli_data")?;
+                return pauli_data
+                    .try_iter()?
+                    .map(|pauli| py_pack_param(&pauli?, qpy_data, Endian::Little))
+                    .collect::<PyResult<_>>();
+            }
+        }
+
+        if let OperationRef::Unitary(unitary) = instruction.op.view() {
+            // unitary gates are special since they are uniquely determined by a matrix, which is not
+            // a "parameter", strictly speaking, but is treated as such when serializing
+
+            // until we change the QPY version or verify we get the exact same result,
+            // we translate the matrix to numpy and then serialize it like python does
+            let out_array = match &unitary.array {
+                ArrayType::NDArray(arr) => arr.to_pyarray(py),
+                ArrayType::OneQ(arr) => arr.to_pyarray(py),
+                ArrayType::TwoQ(arr) => arr.to_pyarray(py),
+            };
+            return Ok(vec![py_pack_param(&out_array, qpy_data, Endian::Little)?]);
+        }
+        if let OperationRef::PauliProductMeasurement(pauli_product_measurement) =
+            instruction.op.view()
+        {
+            let z_values = GenericValue::Tuple(
+                pauli_product_measurement
+                    .z
+                    .iter()
+                    .cloned()
+                    .map(GenericValue::Bool)
+                    .collect(),
+            );
+            let x_values = GenericValue::Tuple(
+                pauli_product_measurement
+                    .x
+                    .iter()
+                    .cloned()
+                    .map(GenericValue::Bool)
+                    .collect(),
+            );
+            let neg_value = GenericValue::Bool(pauli_product_measurement.neg);
+            return Ok(vec![
+                pack_generic_value(&z_values, qpy_data)?,
+                pack_generic_value(&x_values, qpy_data)?,
+                pack_generic_value(&neg_value, qpy_data)?,
+            ]);
+        }
+        instruction
+            .params_view()
+            .iter()
+            .map(|x| pack_param_obj(x, qpy_data, Endian::Little))
+            .collect::<PyResult<_>>()
+    })
```

Comment:
These branches do not require a Python token. If you adopt the suggestion here: https://github.com/Qiskit/qiskit/pull/14166/files#r2573807031 - maybe those parts can move to `pack_instruction` or with maybe a Python-independent function for handling `OperationRef::PauliProductMeasurement`.